### PR TITLE
[Integrity Checker] Implemented pairwise Update normalizer

### PR DESCRIPTION
**Pull request by tudor-da**
This PR enriches the state update comparison in the `IntegrityChecker` with a pairwise update normalizer, meant to adapt the actual/expected updates in relation with their counterpart
  
CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/UpdateNormalizer.scala [hunk 1]
```diff
@@ -29,6 +29,26 @@ object UpdateNormalizer {
...
+object PairwiseUpdateNormalizer {
+  def normalize(
+      expectedUpdate: Update,
+      actualUpdate: Update,
+      normalizers: Iterable[PairwiseUpdateNormalizer],
+  ): (Update, Update) = {
+    normalizers.foldLeft(expectedUpdate -> actualUpdate) { case ((expected, actual), normalizer) =>
+      normalizer.normalize(expected, actual)
+    }
+  }
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/UpdateNormalizer.scala:49]: You can remove the body's curly braces here.

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/UpdateNormalizer.scala [hunk 2]
```diff
@@ -29,6 +29,26 @@ object UpdateNormalizer {
   }
 }
 
+/** Normalizes the expected and/or actual updates based on their counterparts.
+  * Useful when scalar update normalization is not possible.
+  * (e.g. ensure contents of one update are a superset of the ones in the other update)
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/UpdateNormalizer.scala:34]:
```suggestion
  * Useful when independent update normalization is not possible
  * (e.g. ensuring that the content of the actual update is a super-set of the content of the expected update).
```

File: ledger/participant-state/kvutils/tools/src/test/scala/ledger/participant/state/kvutils/tools/integritycheck/StateUpdateComparisonSpec.scala [hunk 3]
```diff
@@ -31,7 +31,7 @@ final class StateUpdateComparisonSpec
       val right = aConfigurationChangeRejected.copy(rejectionReason = "another reason")
 
       ReadServiceStateUpdateComparison
-        .compareUpdates(left, right, List.empty, List.empty)
+        .compareUpdates(left, right, List.empty, List.empty, List.empty)
```
Comment (fabiotudone-da) [inline ledger/participant-state/kvutils/tools/src/test/scala/ledger/participant/state/kvutils/tools/integritycheck/StateUpdateComparisonSpec.scala:34]:
```suggestion
        .compareUpdates(left, right, Iterable.empty, Iterable.empty, Iterable.empty)
```

for coherence with the usage below

File: ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/UpdateNormalizer.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/UpdateNormalizer.scala]: Done

**Comment by fabiotudone-da** [comment]
> The code that is in looks good to me. However, there is no user for it right now. Will that be added in this or another PR?

AFAIK this is code infrastructure that may be used in the future, here or in integration-specific integrity checks.

**Review by fabiotudone-da — APPROVED** [review]

**Comment by tudor-da** [comment]
> The code that is in looks good to me. However, there is no user for it right now. Will that be added in this or another PR?

The core implementation of the integrity checker in the DAML SDK. For future uses (i.e. when tests asserting historical exports are added), it's better to have it here. Furthermore, the state comparison and normalization logic is in here as well, so the small change in here offloads for bigger efforts in users of this core lib.

**Pull request merged**